jerry-core/ecma/base/ecma-gc.c

```diff
@@ -786,7 +786,11 @@ ecma_gc_mark (ecma_object_t *object_p) /**< object to mark from */
       case ECMA_OBJECT_TYPE_PROXY:
       {
         ecma_gc_mark_proxy_object (object_p);
-        break;
+        /* No need to free the properties of a proxy (there should be none) */
+        JERRY_ASSERT ((object_p->u1.property_list_cp == JMEM_CP_NULL)
+                      || (ECMA_GET_FIRST_BIT_FROM_POINTER_TAG (object_p->u1.property_list_cp) != 0)
+                      || (ECMA_GET_SECOND_BIT_FROM_POINTER_TAG (object_p->u1.property_list_cp) != 0));
```

Comment:
The `(ECMA_GET_FIRST_BIT_FROM_POINTER_TAG (object_p->u1.property_list_cp) != 0` condition invalidates the whole point of `(object_p->u1.property_list_cp == JMEM_CP_NULL)`, doesn't it?

The condition should look like: `(object_p->u1.property_list_cp & ~0x3) == JMEM_CP_NULL` (of course with some macros)

Reply:
The reason I separated into three different case is to clearly indicate what are the allowed cases/situations:
1) The bit 1 and 2 is not set
2) The bit 1 is set
3) The bit 2 is set (this also implies bit 1, maybe the assert order is incorrect)


Should I change it to the `~0x3` format? (Is there a macro for the 0x3 "bits"?)

Reply:
I think you misunderstand me. It seems to me you want to test here that property_list_cp is null except the last two bits. However, the assert is wrong in that case, becasue the second or third case invalidates the first case.

Reply:
I.e. 0x101 is accepted by the assert, while it should not be.

Reply:
Since currently only the first and the second pointer tags are used both assertions are correct. However I more like @zherczeg suggestion since it catches any misbehavior upon the third pointer tag is set by accident.

Reply:
Ok Will update, but that is not part of this PR, but the #4264 which this PR depends on.

Reply:
I may misunderstand something, but this is a logical or operation. The first part check if the pointer is null, the second part checks whether the fisrt bit is set. The problem is, if the second part is true, it doesn't matter the first part. Based on your explanation, 3 values are valid: 0x0, 0x1, 0x3. But the assert accepts 0x101, which is not allowed. So the assert is wrong.